Code/Tools/AssetProcessor/native/utilities/UuidManager.h

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#pragma once
+
+#include <AzCore/IO/Path/Path.h>
+#include <AzCore/Interface/Interface.h>
+#include <AzCore/std/containers/unordered_set.h>
+#include <AzCore/std/string/string.h>
+#include <native/AssetManager/SourceAssetReference.h>
+
+namespace AzToolsFramework
+{
+    struct IMetadataRequests;
+}
+
+namespace AssetProcessor
+{
+    struct IUuidRequests
+    {
+        AZ_RTTI(IUuidRequests, "{4EA7E0F6-CB4E-4F9C-ADBC-807676D51772}");
+
+        virtual ~IUuidRequests() = default;
+
+        //! Gets the canonical UUID for a given source asset.  A new metadata file may be created with a randomly generated UUID if
+        //! generation is enabled for the file type.  Otherwise the UUID returned will follow the legacy system based on the file path.
+        virtual AZ::Uuid GetUuid(const SourceAssetReference& sourceAsset) = 0;
```

Comment:
This API seems to only take in absolute paths to source assets, will there be a way to request UUIDs using relative paths to source assets? 

Possible API: "GetScanFoldersForRelativeSourceAssetPath(path)" or something?

Reply:
I don't believe so, no.  This API is AP-specific right now (not sure yet what we'll be exposing to AzToolsFramework), so there's no need for that.  But we generally want to move away from accepting relative paths (by themselves) for assets, since without a known scanfolder, that involves the override system which is usually not intentional.  The override system isn't compatible with the metadata system anyway (for a given asset type).

Is there a usecase you're concerned about?

Reply:
right now the procedural prefab system is referring to the assets (AzToolsFramework::Prefab::PrefabDomUtils::Internal::FixUpInvalidAssets) by hints to fix up the asset IDs 

As soon as the asset sub-id becomes {source UUID}."my_model.azmodel" then this will become moot, but the use of relative paths will need to persist for the foreseeable future; 

That being mentioned, the system could use a "go through the scan folders + this relative path" to make candidate SourceAssetReference sourceAsset or something;

Reply:
Ok, that should be fine as-is.  This is all AP internal code and that fixup happens in AssetBuilder.  This is basically adding support for replacing `CreateSafeSourceUUIDFromName` calls inside AP